### ISSUE TITLE
Restart song when previous track button is pressed if current song position is greater than 15 seconds

### ISSUE THREAD

**Issue by superlgn** [multani/sonata#101]
Hey all. Long time Sonata user here (belated Thanks!). I realize Sonata development is in limbo but I thought I'd file a feature request here regardless so it gets documented if/when a new maintainer ever comes along.

One minor issue that's always bothered me is when I want to restart a song I have to either hit stop+play or try to get the start of the track with the progress bar, which I rarely manage to hit right at 0. Seems to me that most music players have a certain threshold that causes them to restart playback instead of going to the previous, like 5+ or a configurable amount of seconds. I'd love to see this myself.

I've written up a patch with a hard coded value just to kick things off. It does work in my brief testing, but I should mention that I'm completely unfamiliar with Sonata's code and never really did anything with Python prior to this:

```
--- sonata/main.py.orig	2019-01-15 05:56:10.000000000 -0600
+++ sonata/main.py	2019-07-25 15:46:45.225581167 -0500
@@ -2501,9 +2501,13 @@
             self.iterate_now()
 
     def mpd_prev(self, _widget, _key=None):
-        if self.conn:
-            self.mpd.previous()
-            self.iterate_now()
+        if self.conn and self.status_is_play_or_pause():
+                at, length = [int(c) for c in self.status['time'].split(':')]
+                if at >= 15:
+                    self.seek(int(self.status['song']), 0)
+                else:
+                    self.mpd.previous()
+                    self.iterate_now()
 
     def mpd_next(self, _widget, _key=None):
         if self.conn:
```

**Comment by multani** [comment]
Do you have an example of a MPD client which implements a behavior like that? We could compare the implementation then.

**Comment by superlgn** [comment]
Unfortunately, no. Sonata is really the only MPD client I've used since I switched to MPD some years back. I looked at the source for gmpc, glurp, gbmol, and ario, and they all use mpd_player_prev() from libmpd, mpd_sendPrevCommand from libmpdclient, or otherwise send the 'previous' command, all of which appear to result in a track change.

My familiarity with this previous track behavior is really with other players like Xmms (X), Rockbox (firmware for old mp3 devices), Google Play Music, Pulsar (Android), etc.

Wikipedia doesn't specifically detail this exact behavior on their media controls article <https://en.wikipedia.org/wiki/Media_control_symbols>, but they do say the |<< button can double as a skip to the start *or* previous file/track/chapter, and I think that's really what they're alluding to.

The closest thing I could find for MPD is an old post on the Ubuntu forums <https://ubuntuforums.org/showthread.php?t=2004841> where someone wrote up a shell script to create this behavior for the command line:

```
test=$(mpc | grep -Eo '[0-9]+:[0-9][0-9]/' | sed 's|\([0-9].*\):\([0-9][0-9]\)/|\1\2|')
if [ "$test" -ge 5 ]; then
 mpc seek 0
else
 mpc prev
fi
```

I'm pretty sure Rockbox and Google Play Music do this at 5 seconds. Pulsar definitely does, and has no tunables for it. As this behavior goes, 15 seconds is probably going to be on the higher side of things. 5 could be a whole lot more common.

**Comment by multani** [comment]
@superlgn Thanks for the thorough search!
In my experience with other clients as well, I think 5 seconds is something more common than 15 (I think even 5 may be a bit on the higher end, but well.)

Would you mind opening a PR for this change? It's probably mergeable more or less like this. 2 changes that you could do beforehand:

* move the 15 magic number into the `sonata/consts.py` file, give it a proper name and update the value to 5 ;)
* your original patch change the logic of the prev command a bit and it looks like you won't be able to go to the previous song if the player is ***not** in play or pause mode. Can you change this back as well?

Thanks again!

**Comment by superlgn** [comment]
Yeah, I paused for a moment when I made the logic change you're speaking of. But when I saw that the previous and next track buttons didn't work unless a song was already playing anyway ([MainThread] sonata.mpdhelper: [55@0] {} Not playing) I went ahead with the changes thinking it'd be less duplicate code and that I'd fix a warning in the process. I didn't really think anything of it until you brought it up now. That's my bad.

Sounds like this is not the intended behavior. I don't know if it's a bug with Sonata or perhaps something with my version of MPD (Debian 10 mpd-0.21.5-3 package). But I did restore the old logic in my pull request (pull #102).

It's my first pull request. Hopefully everything is ok.

Thanks,

Mike
